File: c/graph/maximum_flow/HLPP.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct node {
    int index;
    int weight;
    struct node *next;
} EdgeNode;

typedef struct {
    int value;
    EdgeNode *next;
} VertexNode;

typedef VertexNode *pNode;

typedef struct {
    int index;
    int degree;
} HeapNode;

typedef struct {
    HeapNode *values;
    int length;
} QueueNode;

static void search(pNode *graphs, int length, int start, int end, int **weights, int *gap, int *degrees, QueueNode *queue, int *result);
static int min(int i, int j);
static QueueNode *createQueue(int length);
static void enqueue(QueueNode *queue, int index, int weight);
static HeapNode dequeue(QueueNode *queue);

int hlpp(pNode *graphs, int length, int start, int end) {
    int **weights = (int **) malloc(sizeof(int *) * length);
    for (int i = 0; i < length; ++i) {
        weights[i] = (int *) malloc(sizeof(int) * length);
        memset(weights[i], 0, sizeof(int) * length);
        EdgeNode *edge = graphs[i]->next;
        while (edge) {
            weights[i][edge->index] = edge->weight;
            edge = edge->next;
        }
    }

    int *gap = (int *) malloc(sizeof(int) * length);
    memset(gap, 0, sizeof(int) * length);
    int *degrees = (int *) malloc(sizeof(int) * length);
    memset(degrees, 0, sizeof(int) * length);
    degrees[start] = length;
    QueueNode *queue = createQueue(length);
    int *result = (int *) malloc(sizeof(int) * length);
    memset(result, 0, sizeof(int) * length);
    result[start] = ~(1 << 31);

    EdgeNode *edge = graphs[start]->next;
    while (edge) {
        int edgeIndex = edge->index;
        int value = min(result[start], weights[start][edgeIndex]);
        weights[start][edgeIndex] -= value;
        weights[edgeIndex][start] += value;
        result[start] -= value;
        result[edgeIndex] += value;
        if (value && edgeIndex != end) {
            enqueue(queue, edgeIndex, degrees[edgeIndex]);
        }
        edge = edge->next;
    }

    while (queue->length) {
        HeapNode heap = dequeue(queue);
        int index = heap.index;
        if (result[index]) {
            search(graphs, length, index, end, weights, gap, degrees, queue, result);
        }
    }
    return result[end];
}

static void search(pNode *graphs, int length, int start, int end, int **weights, int *gap, int *degrees, QueueNode *queue, int *result) {
    EdgeNode *edge = graphs[start]->next;
    while (edge) {
        int edgeIndex = edge->index;
        if (degrees[start] == degrees[edgeIndex] + 1) {
            int value = min(result[start], weights[start][edgeIndex]);
            weights[start][edgeIndex] -= value;
            weights[edgeIndex][start] += value;
            result[start] -= value;
            result[edgeIndex] += value;
            if (value && edgeIndex != end) {
                enqueue(queue, edgeIndex, degrees[edgeIndex]);
            }
        }
        edge = edge->next;
    }

    if (start != end && result[start]) {
        --gap[degrees[start]];
        if (!~gap[degrees[start]]) {
            for (int i = 0;i < length; ++i) {
                if (i != end && degrees[i] > degrees[start] && degrees[i] < length) {
                    degrees[i]= length;
                }
            }
        }
        ++degrees[start];
        if (degrees[start] < length) {
            ++gap[degrees[start]];
            enqueue(queue, start, degrees[start]);
        }
    }
}

static int min(int i, int j) {
    int aux = ~((i > j) ? 1 : 0) + 1;
    return (i & ~aux) | (j & aux);
}

static QueueNode *createQueue(int length) {
    QueueNode *queue = (QueueNode *) malloc(sizeof(QueueNode));
    queue->values = (HeapNode *) malloc(sizeof(HeapNode) * length);
    queue->length = 0;
    return queue;
}

static void enqueue(QueueNode *queue, int index, int degree) {
    int child = queue->length;
    while (child > 0) {
        int parent = (child - 1) >> 1;
        if (queue->values[parent].degree >= degree) {
            break;
        }
        queue->values[child] = queue->values[parent];
        child = parent;
    }
    queue->values[child].index = index;
    queue->values[child].degree = degree;
    ++queue->length;
}

static HeapNode dequeue(QueueNode *queue) {
    --queue->length;
    HeapNode value = queue->values[0];
    HeapNode pivot = queue->values[queue->length];
    int parent = 0;
    while (1) {
        int child = (parent << 1) + 1;
        if (child >= queue->length) {
            break;
        }

        if (child < queue->length - 1 && queue->values[child].degree < queue->values[child + 1].degree) {
            ++child;
        }
        if (queue->values[child].degree <= pivot.degree) {
            break;
        }
        queue->values[parent] = queue->values[child];
        parent = child;
    }
    queue->values[parent] = pivot;
    return value;
}

File: java/graph/maximum_flow/HLPP.java
package graph.maximum_flow;

import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Queue;

public class HLPP {
    public int search(VertexNode[] graphs, int start, int end) {
        int[][] weights = new int[graphs.length][graphs.length];
        for (int i = 0; i < graphs.length; ++i) {
            Arrays.fill(weights[i], 0);
            EdgeNode edge = graphs[i].next;
            while (null != edge) {
                weights[i][edge.index] = edge.weight;
                edge = edge.next;
            }
        }

        int[] gap = new int[graphs.length];
        Arrays.fill(gap, 0);
        int[] degrees = new int[graphs.length];
        Arrays.fill(degrees, 0);
        degrees[start] = graphs.length;
        Queue<HeapNode> queue = new PriorityQueue<>(cmp);
        int[] result = new int[graphs.length];
        Arrays.fill(result, 0);
        result[start] = Integer.MAX_VALUE;

        EdgeNode edge = graphs[start].next;
        while (null != edge) {
            int edgeIndex = edge.index;
            int value = Math.min(result[start], weights[start][edgeIndex]);
            weights[start][edgeIndex] -= value;
            weights[edgeIndex][start] += value;
            result[start] -= value;
            result[edgeIndex] += value;
            if (0 != value && edgeIndex != end) {
                queue.offer(new HeapNode(edgeIndex, degrees[edgeIndex]));
            }
            edge = edge.next;
        }

        while (!queue.isEmpty()) {
            HeapNode heap = queue.poll();
            int index = heap.index;
            if (0 != result[index]) {
                search(graphs, index, end, weights, gap, degrees, queue, result);
            }
        }
        return result[end];
    }

    private void search(VertexNode[] graphs, int start, int end, int[][] weights, int[] gap, int[] degrees, Queue<HeapNode> queue, int[] result) {
        EdgeNode edge = graphs[start].next;
        while (null != edge) {
            int edgeIndex = edge.index;
            if (degrees[start] == degrees[edgeIndex] + 1) {
                int value = Math.min(result[start], weights[start][edgeIndex]);
                weights[start][edgeIndex] -= value;
                weights[edgeIndex][start] += value;
                result[start] -= value;
                result[edgeIndex] += value;
                if (0 != value && edgeIndex != end) {
                    queue.offer(new HeapNode(edgeIndex, degrees[edgeIndex]));
                }
            }
            edge = edge.next;
        }

        if (start != end && 0 != result[start]) {
            --gap[degrees[start]];
            if (0 == ~gap[degrees[start]]) {
                for (int i = 0; i < graphs.length; ++i) {
                    if (i != end && degrees[i] > degrees[start] && degrees[i] < graphs.length) {
                        degrees[i]= graphs.length;
                    }
                }
            }
            ++degrees[start];
            if (degrees[start] < graphs.length) {
                ++gap[degrees[start]];
                queue.offer(new HeapNode(start, degrees[start]));
            }
        }
    }

    private class HeapNode {
        public int index = 0;
        public int degree = 0;

        public HeapNode(int index, int degree) {
            this.index = index;
            this.degree = degree;
        }
    }

    private static Comparator<HeapNode> cmp = new Comparator<HeapNode>() {
        @Override
        public int compare(HeapNode node1, HeapNode node2) {
            return node2.degree - node1.degree;
        }
    };

    public class EdgeNode {
        public int index = 0;
        public int weight = 0;
        public EdgeNode next = null;

        public EdgeNode(int index, int weight) {
            this.index = index;
            this.weight = weight;
        }
    }

    public class VertexNode {
        public int value = 0;
        public EdgeNode next = null;

        public VertexNode(int value) {
            this.value = value;
        }
    }
}
